src/textual/widgets/text_input.py

```diff
@@ -0,0 +1,172 @@
+            self.cursor_index = min(len(self.current_text), self.cursor_index + 1)
+            changed = True
+
+        if changed:
+            self.post_message_no_wait(self.Changed(self, value=self.current_text))
+
+
+class TextInput(TextInputBase, can_focus=True):
+    CSS = """
+    TextInput {
+        width: auto;
+        background: $primary;
+        height: 3;
+        padding: 0 1;
+        content-align: left middle;
+    }
+
+    TextInput:hover {
+        background: $primary-darken-1;
+    }
+
+    TextInput:focus {
+        background: $primary-darken-2;
+        border: solid $primary-lighten-1;
+        padding: 0;
+    }
+
+    App.-show-focus TextInput:focus {
+        tint: $accent 20%;
+    }
+    """
+
+    class Submitted(Message, bubble=True):
+        def __init__(self, sender: MessageTarget, value: str) -> None:
+            """Message posted when the user presses the 'enter' key while
+            focused on a TextInput widget.
+
+            Args:
+                sender (MessageTarget): Sender of the message
+                value (str): The value in the TextInput
+            """
+            super().__init__(sender)
+            self.value = value
+
+    def __init__(
+        self,
+        placeholder: str = "",
+        initial: str = "",
+        *,
+        name: str | None = None,
+        id: str | None = None,
+        classes: str | None = None,
+    ):
+        super().__init__(name=name, id=id, classes=classes)
+        self.placeholder = placeholder
+        self.current_text = initial if initial else ""
+
+    def get_content_width(self, container_size: Size, viewport_size: Size) -> int:
+        return (
+            max(len(self.current_text), len(self.placeholder))
+            + self.styles.gutter.width
+        )
+
+    def render(self) -> RenderableType:
+        # We only show the cursor if the widget has focus
+        show_cursor = self.has_focus
+        if self.current_text:
+            display_text = Text(self.current_text, no_wrap=True)
+
+            if show_cursor:
+                display_text = self._apply_cursor_to_text(
+                    display_text, self.cursor_index
+                )
+            return display_text
+        else:
+            # The user has not entered text - show the placeholder
+            display_text = Text(self.placeholder, "dim")
+            if show_cursor:
+                display_text = self._apply_cursor_to_text(display_text, 0)
+            return display_text
+
+    def _apply_cursor_to_text(self, display_text: Text, index: int):
+        # Either write a cursor character or apply reverse style to cursor location
+        if index == len(display_text):
+            display_text += "█"
+        else:
+            display_text.stylize(
+                "reverse",
+                start=index,
+                end=index + 1,
+            )
+        return display_text
+
+    def on_focus(self, event: events.Focus) -> None:
```

Comment:
I can't remember why I added this or if it's needed 🤷

Reply:
Seems like the focus styles aren't being applied when the widget gets focus unless I manually refresh like this.